pisound-btn/pisound-btn.c

```diff
@@ -233,94 +232,145 @@ static void read_config_value(const char *conf, const char *value_name, char *ds
+static void get_action_script(enum action_e action, char* action_name, char *buffer, char* args, unsigned int buff_length)
 {
-	if (action < 0 || action >= A_COUNT)
-		return -EINVAL;
+	if (action_name[0] == '\0')
+	{
+		buffer[0] = '\0';
+	}
 
-	const char *action_name = NULL;
-	const char *default_script = NULL;
+	read_config_value(g_config_path, action_name, buffer, args, buff_length, "#");
+	// if buffer[0] == '0' then the entry did not exist in the config file.
+	if (buffer[0] == '\0')// No value was specified.
+	{
+		bool found = false;
+		switch (action)
+		{
+		case A_UP:
+			strcpy( buffer, DEFAULT_UP );
+			if (args)
+				args[0] = '\0';
+			break;
+		case A_DOWN:
+			strcpy( buffer, DEFAULT_DOWN );
+			if (args)
+				args[0] = '\0';
+			break;
+		case A_CLICK:
+			if (strcmp("CLICK_1", action_name) == 0)
+			{
+				strcpy( buffer, DEFAULT_CLICK_1 );
+				if (args)
+					args[0] = '\0';
+			}
+			else if (strcmp("CLICK_2", action_name) == 0)
+			{
+				strcpy( buffer, DEFAULT_CLICK_2 );
+				if (args)
+					args[0] = '\0';
+			}
+			else if (strcmp("CLICK_3", action_name) == 0)
+			{
+				strcpy( buffer, DEFAULT_CLICK_3 );
+				if (args)
+					args[0] = '\0';
+			}
+			else
+			{
+				read_config_value(g_config_path, CLICK_OTHER_VALUE_NAME, buffer, args, buff_length, "#");
+			}
+			break;
+		case A_HOLD:
+			if (strcmp("HOLD_3S", action_name) == 0)
+			{
+				strcpy( buffer, DEFAULT_HOLD_3S );
+				if (args)
+					args[0] = '\0';
+			}
+			else if (strcmp("HOLD_5S", action_name) == 0)
+			{
+				strcpy( buffer, DEFAULT_HOLD_5S );
+				if (args)
+					args[0] = '\0';
+			} else {
+				read_config_value(g_config_path, HOLD_OTHER_VALUE_NAME, buffer, args, buff_length, "#");
+			}
+			break;
+		default:
+			break;
+		}
+	}
+	// if there is a # at this point entry existed in the config file but had no value
+	if (buffer[0] == '#' ) {
+		buffer[0] = '\0';
+	}
+}
 
-	get_default_action_and_script(action, arg0, arg1, &action_name, &default_script);
+// 0 or an error (negative number).
+static int get_action_script_path(enum action_e action, char * action_name, unsigned click_count, unsigned hold_time, char *script, char *args, size_t n)
+{
+	if (action < 0 || action >= A_COUNT) {
+		return -EINVAL;
+	}
 
-	if (action_name == NULL || default_script == NULL)
+	get_action_name(action, action_name, click_count, hold_time);
+	if (action_name[0] == '\0')
+	{
 		return -EINVAL;
+	}
 
-	char script[MAX_PATH_LENGTH + 1];
+	char action_script[MAX_PATH_LENGTH + 1];
 	char arguments[MAX_PATH_LENGTH + 1];
 
-	read_config_value(g_config_path, action_name, script, arguments, MAX_PATH_LENGTH, default_script);
+	get_action_script(action, action_name, action_script, arguments, MAX_PATH_LENGTH );
 
-	fprintf(stderr, "script = '%s', args = '%s'\n", script, arguments);
+	fprintf(stderr, "script = '%s', args = '%s'\n", action_script, arguments);
 
+	if (action_script[0] == '\0')
+	{
+		return 0;
```

Comment:
It should set at least script[0] to '\0', to avoid having uninitialized output variable